### ISSUE TITLE
[HOLD for payment 2022-05-20] [$1000] Workspace - Unable delete the unlimited numbers under rate in Reimburse expenses page

### ISSUE THREAD

**Issue by kavimuru** [Expensify/App#8309]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. Go to https://staging.new.expensify.com/
2. Log in with expensifail account
3. Go to Setting - Workspace - Reimburse expenses
4. Enter unlimited numbers and unlimited numbers after point
5. Try to delete the numbers

## Expected Result:
Limiting rate to 8 digits (before the decimal). We should still maintain 3 decimal places because that's the default IRS rate. E.g 12345678.123

## Actual Result:
Unable to delete the numbers under rate in Reimburse expenses page

## Workaround:
Unknown

## Platform:

Where is this issue occurring?

- Web
- iOS
- Android
- Desktop App
- Mobile Web

**Version Number:** 1.1.46.0

**Reproducible in staging?:** Y

**Reproducible in production?:** Y

**Email or phone of affected tester (no customers):** applausetester+0707abb@applause.expensifail.com / Feya87Katya

**Logs:** https://stackoverflow.com/c/expensify/questions/4856

**Notes/Photos/Videos:** 

https://user-images.githubusercontent.com/43996225/160024978-c63139ff-6b1b-479f-87ac-3ac947fd37aa.mp4

**Expensify/Expensify Issue URL:** 

**Issue reported by:** Applause

**Slack conversation:** 

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

**Comment by parasharrajat** [comment]
Lol. This is a good one. We didn't notice that the browser will use the `e+` scientific number format.

**Comment by Gonals** [comment]
Looks like a pretty good issue! Setting as external and weekly

**Comment by kadiealexander** [comment]
Posted to Upwork!

Internal: https://www.upwork.com/ab/applicants/1508929160732057600/job-details
External: https://www.upwork.com/jobs/~0170d8f13e30bba89f

**Comment by parasharrajat** [comment]
This is missing Exported label.

**Comment by kadiealexander** [comment]
Adding @jliexpensify to keep an eye on this while I'm away for the next few days, in case we need to double the price on Thursday.

**Comment by kadiealexander** [comment]
Price doubled!!

**Comment by metehanozyurt** [comment]
## **Proposal**

### Cause
It checks the number and then switches it to the scientific number. Therefore, since the previous state and the current state are scientific numbers, it does not perceive the change.


### Solution
If we parse the number first, we can do the conversion to scientific digits before checking.

we should change this line
https://github.com/Expensify/App/blob/afd220110a33886b95923fe004556a3ca9396261/src/pages/workspace/reimburse/WorkspaceReimburseNoVBAView.js#L85

like this:
```js
const isInvalidRateValue = value !== '' && !CONST.REGEX.RATE_VALUE.test(parseFloat(value));
```

after that we can detect scientific numbers.

video:

https://user-images.githubusercontent.com/6093207/162609474-6d04a1e3-8017-4d8e-a21b-20c50fdf20fe.mov

**Comment by pecanoro** [comment]
Not overdue.

**Comment by kadiealexander** [comment]
Price doubled again!

**Comment by izhan** [comment]
# Proposal

## Cause
Oi these are some big numbers! There are a few issues here:

1. The `Number.prototype.toString()` method will display large numbers as scientific notation. If you're interested in the algo, it's [here](https://tc39.es/ecma262/multipage/ecmascript-data-types-and-values.html#sec-numeric-types-number-tostring); it's a bit detailed, but the main takeaway from that is it starts doing scientific notation for e+21 and up
2. Javascript uses double-precision 64-bit IEEE 754 to represent its [numbers](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number), meaning we'll lose some precision when we call `parseFloat` on it.

Easy way to test yourself in console:
```js
const scientificNotation = 1000000000000000000000
console.log(scientificNotation.toString())
> 1e+21

const losingPrecision = 1234567890123456789
console.log(losingPrecision.toString())
> 1234567890123456800
```

(1) is root cause here. (2) is just another bug I noticed that I think we should fix anyways - you probably don't want to lose precision for user input

## Solution
TLDR; avoid converting to floats; keep precisions in strings

We call `parseFloat` on our input so we can round to nearest thousandth (i.e. the calls to `toFixed(3)`). We lose precision then (which also leads to weird scientific notation), so instead let's keep everything in strings.

We already know that the strings are in proper float format (e.g. XXX.XXX) here https://github.com/Expensify/App/blob/main/src/pages/workspace/reimburse/WorkspaceReimburseNoVBAView.js#L85. So instead of `parseFloat(...)` -> `toFixed(3)`, let's just do some string manipulation. A helper function like this would do the trick:

```js
function truncateToThousandth(str) {
    const num = parseFloat(str);
    if (_.isNaN(num)) {
        return '';
    }
    
    const [whole, decimal] = str.split('.')
    const truncatedDecimal = decimal ? decimal.substring(0, 3).padEnd(3, 0) : '000';
    return `${whole}.${truncatedDecimal}`;
}

truncateToThousandth('12345678901234567890123456789.12345')
> 12345678901234567890123456789.123
```

Note: small regression as it truncates and doesn't round to nearest thousandth, as I figured from the UX perspective it doesn't matter, but we can also make it round w/ a few extra lines of code too. LMK if you prefer that instead.

**Comment by parasharrajat** [comment]
>  figured from the UX perspective it doesn't matter, but we can also make it round

Sorry but it matters. 

I am curious that we probably don't need that big number. What if we limit that input to a max number. @trjExpensify What do you think we should do here? Tagging you as we recently discussed this.

- Should we limit the input value to a max number? I don't think that an arbitrary long number is expected in this field e.g. 12345678901234567890123456789.12345

**Comment by trjExpensify** [comment]
>Should we limit the input value to a max number? I don't think that an arbitrarily long number is expected in this field e.g. 12345678901234567890123456789.12345

Yeah, good question! @pecanoro @iwiznia, is there a max number of digits we allow on OldDot for the mileage rate input field? We could align with that, or perhaps use the same limit as `Request Money` which is set at `8`? I tend to agree that an arbitrary long number isn't really required for mileage, who's reimbursing millions per mile travelled?.. 😅

**Comment by metehanozyurt** [comment]
I think the problem here is because the number is converted to scientific number. If the digits that can check before transform. Thus, when the user is prevented from entering a new number than he can delete the number again. That's what I meant in my previous offer. If it can be prevented from turning into a scientific number, it rounds the number after the dot with 3 digits in the current approach.
I liked my offer a lot earlier, but I wanted to try say it again because of my concerns that I couldn't explain it. 🙈

**Comment by iwiznia** [comment]
@trjExpensify I just tested (in app, you can do it too) with 12345678901234567890 and the value it saved was 12,345,678,901,234,567,000.00, so there seems to be a limit, but I bet that limit is the max int value or something like that.
Anyway, yes, let's add some arbitrary limit if it is causing problems.

**Comment by trjExpensify** [comment]
Cool, so I'd vote we go for limiting to 8 digits (before the decimal) like `Request money`. We still need to maintain `3` decimal places because that's the default IRS rate. E.g `12345678.123`. Thoughts?

**Comment by izhan** [comment]
Yeah anything less than `Number.MAX_INT` (which is 9007199254740991) works, so 8 digits is reasonable! Best place IMO is putting it in the regex.

Out of curiosity - is there a reason we are correcting the value in the debounce function, and not in `setRate` when we set it? We already enforce the format on input using the regex, so as a user I was surprised that the UI formatted my input a 2nd time after a delay. Would it be better UX if we just consolidate all formatting logic to be on input only?

**Comment by parasharrajat** [comment]
I am good with https://github.com/Expensify/App/issues/8309#issuecomment-1105362009. What do you think @pecanoro? 

>  is there a reason we are correcting the value in the debounce function, and not in setRate when we set it?

So that users can type freely and modify the value if needed.

**Comment by pecanoro** [comment]
Yes! It sounds good, I doubt the rate will go to anything closer to that ever, or at least I doubt we will be here to see it hahaha

**Comment by parasharrajat** [comment]
Can someone update the Expected behavior with https://github.com/Expensify/App/issues/8309#issuecomment-1105362009? 

Accepting proposals now for the new [behavior](https://github.com/Expensify/App/issues/8309#issuecomment-1105362009)......

**Comment by izhan** [comment]
I'm somewhat new to this – I imagine I can just resubmit a proposal w/ a solution only (given I explained the problem above)?

# Proposal

### Solution
Change the regex we use in `setRate` to only limit to 8 characters. So changing this line: https://github.com/Expensify/App/blob/7029b6fdb8d1bb0617ffa695ef05fa52de316f10/src/CONST.js#L633

to `/^\d{1,8}+(\.\d*)?$/`

This has the intended behavior, as shown below
![Screen Shot 2022-04-21 at 1 02 06 PM](https://user-images.githubusercontent.com/4621962/164546478-dec773c8-8a89-4c9c-a99e-b0ff40270994.png)

Quick UX note: this makes the behavior more in line with that of `Request Money`, i.e. users cannot enter > 8 digits. We could alternatively add this check AFTER the setState (when we sent it to the server), but that leads to funky UX questions, i.e. how should we truncate long input? So IMO this is probably best, though open to other ideas

**Comment by parasharrajat** [comment]
I am also confused to determine the best solution for this problem. we recently published some [rules](https://github.com/Expensify/App/blob/main/FORMS.md#modifying-user-input-on-change) which restrict us from manipulating the user input. 

Pulling @luacmartins to determine the best course here. :bow:

**Comment by luacmartins** [comment]
@parasharrajat what part exactly do you need help with? I think it's fine to limit the rate to 8 digits before decimal and 3 decimal digits.

**Comment by parasharrajat** [comment]
Ok, Thanks. I will review the proposal tomorrow.

**Comment by izhan** [comment]
Hey @parasharrajat - no rush, but just wanted to see if there's anything I can provide here!

**Comment by parasharrajat** [comment]
Sorry, I was sick for few days past week. Looking into now.

**Comment by parasharrajat** [comment]
@izhan's proposal is looking good. 

cc: @pecanoro 

:ribbon: :eyes: :ribbon:  C+ reviewed

**Comment by izhan** [comment]
I think that was never an issue (here I'm adding >3 decimals on staging)

https://user-images.githubusercontent.com/4621962/165826793-90a0711b-8460-4ff2-93dd-d38582f5e989.mov

I think the original issue was removing decimals of any large number that converted to scientific notation, which this PR should fix with the 8-digit limit

**Comment by pecanoro** [comment]
Ohh you are right, the scientific notation was the issue here. Perfect, all good then, go ahead and apply for the job!

**Comment by izhan** [comment]
PR is here https://github.com/Expensify/App/pull/8838

**Comment by mvtglobally** [comment]
Issue not reproducible during KI retests. (First week)

**Comment by kadiealexander** [comment]
@izhan - paid!
@parasharrajat - contract sent over, please let me know when you accept!

**Comment by parasharrajat** [comment]
@kadiealexander Done.

**Comment by kadiealexander** [comment]
All paid!! Thanks team!
